includes/sanitizers/class-amp-img-sanitizer.php

```diff
@@ -18,51 +18,85 @@ class AMP_Img_Sanitizer extends AMP_Base_Sanitizer {
 	private static $script_src = 'https://cdn.ampproject.org/v0/amp-anim-0.1.js';
 
 	public function sanitize() {
+
 		$nodes = $this->dom->getElementsByTagName( self::$tag );
+		$have_dimensions = array();
+		$need_dimensions = array();
+
 		$num_nodes = $nodes->length;
+
 		if ( 0 === $num_nodes ) {
 			return;
 		}
 
 		for ( $i = $num_nodes - 1; $i >= 0; $i-- ) {
 			$node = $nodes->item( $i );
-			$old_attributes = AMP_DOM_Utils::get_node_attributes_as_assoc_array( $node );
 
-			if ( empty( $old_attributes['src'] ) ) {
+			if ( ! $node->hasAttribute( 'src' ) || '' === $node->getAttribute( 'src' ) ) {
 				$node->parentNode->removeChild( $node );
 				continue;
 			}
 
-			$new_attributes = $this->filter_attributes( $old_attributes );
-
-			// Try to extract dimensions for the image, if not set.
-			if ( ! isset( $new_attributes['width'] ) || ! isset( $new_attributes['height'] ) ) {
-				$dimensions = AMP_Image_Dimension_Extractor::extract( $new_attributes['src'] );
-				if ( is_array( $dimensions ) ) {
-					$new_attributes['width'] = $dimensions[0];
-					$new_attributes['height'] = $dimensions[1];
-				}
+			// Determine which images need their dimensions determined/extracted.
+			if ( ! $node->hasAttribute( 'width' ) || ! $node->hasAttribute( 'height' ) ) {
+				$need_dimensions[ $node->getAttribute( 'src' ) ][] = $node;
+			} else {
+				$have_dimensions[ $node->getAttribute( 'src' ) ][] = $node;
 			}
+		}
+		$this->determine_dimensions( $need_dimensions );
 
-			// Final fallback when we have no dimensions.
-			if ( ! isset( $new_attributes['width'] ) || ! isset( $new_attributes['height'] ) ) {
-				$new_attributes['width'] = isset( $this->args['content_max_width'] ) ? $this->args['content_max_width'] : self::FALLBACK_WIDTH;
-				$new_attributes['height'] = self::FALLBACK_HEIGHT;
+		$this->adjust_and_replace( $need_dimensions );
+		$this->adjust_and_replace( $have_dimensions );
```

Comment:
So, reading through this a few times, I wonder if we could simplify by doing two separate passes: once for images with dimensions, then fetch any missing dimensions and repeat for remaining images.

Right now, it feels we've optimized for the `needs_dimension` use case and are probably doing multiple unnecessary loops (loop 1: figure out which nodes need dimensions; loop 2+: [nested] process need_dimensions; loop 3+: [nested] process have_dimensions).

We could probably combine 1 and 3 and do both things in the same pass:

```
for ( img_nodes ) {
    if ( has_dimensions ) {
        process_node
    } else {
        array_push( node, remaining_nodes )
    }
}

if ( remaining_nodes ) {
    determine dimensions for remaining_nodes

    for ( remaining_nodes ) {
        process_node
    }
}
```

it's also possible that I may be overlooking something :) Thoughts?